src/provider.lua

```diff
@@ -0,0 +1,306 @@
+  ngx.header.content_type = service.auth_missing_headers
+  ngx.print(service.error_auth_missing)
+  ngx.exit(ngx.HTTP_OK)
+end
+
+local function error_authorization_failed(service)
+  ngx.log(ngx.INFO, 'authorization failed for service ' .. tostring(service.id))
+
+  ngx.status = service.auth_failed_status
+  ngx.header.content_type = service.auth_failed_headers
+  ngx.print(service.error_auth_failed)
+  ngx.exit(ngx.HTTP_OK)
+end
+
+local function error_no_match(service)
+  ngx.log(ngx.INFO, 'no rules matched for service ' .. tostring(service.id))
+  ngx.status = service.no_match_status
+  ngx.header.content_type = service.no_match_headers
+  ngx.print(service.error_no_match)
+  ngx.exit(ngx.HTTP_OK)
+end
+-- End Error Codes
+
+-- Aux function to split a string
+
+function string:split(delimiter)
+  local result = { }
+  local from = 1
+  local delim_from, delim_to = string.find( self, delimiter, from )
+  if delim_from == nil then return {self} end
+  while delim_from do
+    table.insert( result, string.sub( self, from , delim_from-1 ) )
+    from = delim_to + 1
+    delim_from, delim_to = string.find( self, delimiter, from )
+  end
+  table.insert( result, string.sub( self, from ) )
+  return result
+end
+
+local function first_values(a)
+  local r = {}
+  for k,v in pairs(a) do
+    if type(v) == "table" then
+      r[k] = v[1]
+    else
+      r[k] = v
+    end
+  end
+  return r
+end
+
+local function build_querystring_formatter(fmt)
+  return function (query)
+    local function kvmap(f, t)
+      local res = {}
+      for k, v in pairs(t) do
+        table.insert(res, f(k, v))
+      end
+      return res
+    end
+
+    return table.concat(kvmap(function(k,v) return string.format(fmt, k, v) end, query or {}), "&")
+  end
+end
+
+local build_querystring = build_querystring_formatter("usage[%s]=%s")
+local build_query = build_querystring_formatter("%s=%s")
+
+
+--[[
+  Authorization logic
+]]--
+
+local function get_auth_params(where, method)
+  local params = {}
+  if where == "headers" then
+    params = ngx.req.get_headers()
+  elseif method == "GET" then
+    params = ngx.req.get_uri_args()
+  else
+    ngx.req.read_body()
+    params = ngx.req.get_post_args()
+  end
+  return first_values(params)
+end
+
+local function get_debug_value()
+  local h = ngx.req.get_headers()
+  if h["X-3scale-debug"] == configuration.provider_key then
+    return true
+  else
+    return false
+  end
+end
+
+function _M.find_service(host)
+  for _,service in ipairs(_M.services or {}) do
+    for _,_host in ipairs(service.hosts or {}) do
+      if _host == host then
+        return service
+      end
+    end
+  end
+end
+
+local http = {
+  get = function(url)
+    ngx.log(ngx.INFO, '[http] requesting ' .. url)
+    local res = ngx.location.capture(assert(url), { share_all_vars = true })
+
+    ngx.log(ngx.INFO, '[http] status: ' .. tostring(res.status))
+    return res
+  end
+}
+local function oauth(params, service)
```

Comment:
Yeah, there are plenty of improvements. So far it was just a copy from the generated configuration.